### DISCUSSION TITLE
classic mirrored queue，producer and consumer connect to different  node，has diff tps。

### DISCUSSION THREAD

**Discussion by caolicaoli** [rabbitmq/rabbitmq-server#9016] [Other] [answered]
### Describe the bug

classic mirrored queue， producer connect master node，  consumer connect mirror node，the tps is 3500 message， but produc and consumer both connect to masternode， the tps is 17000， how is that ？

### Reproduction steps

classic mirrored queue， producer connect master node，  consumer connect mirror node，the tps is 3500 message， but produc and consumer both connect to masternode， the tps is 17000， how is that ？

### Expected behavior

the tps is almost same。。。。

**Comment by michaelklishin** [comment, marked as answer]
Classic mirrored queues is a feature that has been deprecated and will be [removed from RabbitMQ in early 2024](https://blog.rabbitmq.com/posts/2023/02/quorum-queues-migration/).

It's a data locality effect. Consumers do not consume from mirrors (followers) in the case of classic mirrored queues.

This is a yet another reason to migrate to quorum queues where consumers can get deliveries directly from followers (but acknowledgements still flow through the leader).

**Comment by michaelklishin** [comment]
Data locality matters for all the messaging protocols RabbitMQ supports, for all queue types but specifically for the long deprecated classic mirrored queues where *all* operations, including deliveries, goes via the node that hosts the leader replica.

This is not really the case for streams because clients discover the topology and connect to the nodes where the replicas are. This [has its own downsides](https://blog.rabbitmq.com/posts/2021/07/connecting-to-streams/) but as far as throughput goes, is a helpful protocol design decision.
